File: Atcoder/AGC028/Q2.cpp
#include <iostream>
#include <algorithm>

using ll = long long;

constexpr ll MOD = 1000000007;
constexpr int MAX_N = 100000;

ll N;
ll P[MAX_N];

ll ext_gcd(ll a, ll b, ll &x, ll &y){
	if(b == 0){
		x = 1, y = 0;
		return a;
	}
	ll d = ext_gcd(b, a % b, y, x);
	y -= a / b * x;
	return d;
}

ll inv_mod(ll a, ll m){
	ll x, y;
	ext_gcd(a, m, x, y);
	x %= m;
	if(x < 0) x += m;
	return x;
}

int main(){

	std::cin >> N;

	for(int i = 0; i < N; ++i){
		ll x = inv_mod(i + 1, MOD);
		if(i > 0) P[i] = x;
		else P[i] = P[i - 1] + x;
		std::cout << P[i] % MOD << std::endl;
	}

	ll ans = 0;
	for(int i = 0; i < N; ++i){
		ll input;
		std::cin >> input;
		ans += input * (P[i] + P[N - 1] - P[i]) % MOD;
		ans %= MOD;
	}

	std::cout << ans << std::endl;

	return 0;
}

File: Atcoder/AGC033/Q2.py
H, W, N = map(int, input().split())
sr, sc = map(int, input().split())
S = input()
T = input()

aL = aR = aU = aD = 0
bL = bR = bU = bD = 0

for c in S:
    if c == 'L':
        aL += 1
    elif c == 'R':
        aR += 1
    elif c == 'U':
        aU += 1
    elif c == 'D':
        aD += 1

for c in T:
    if c == 'L':
        bL += 1
    elif c == 'R':
        bR += 1
    elif c == 'U':
        bU += 1
    elif c == 'D':
        bD += 1

cL = sc - 1
cR = W - sc
cU = sr - 1
cD = H - sr

if cL == 0 and aL != 0:
    ans = False
elif cR == 0 and aR != 0:
    ans = False
elif cU == 0 and aU != 0:
    ans = False
elif cD == 0 and aD != 0:
    ans = False

if aL - bL > cL or aR - bR > cR or\
        aU - bU > cU or aD - bD > cD:
    ans = False
else:
    ans = True
